src/rgw/rgw_asio_frontend.cc

```diff
@@ -68,62 +60,136 @@ void Pauser::wait()
 }
 
 using tcp = boost::asio::ip::tcp;
+namespace beast = boost::beast;
 
-// coroutine to handle a client connection to completion
-static void handle_connection(RGWProcessEnv& env, tcp::socket socket,
-                              boost::asio::yield_context yield)
-{
-  auto cct = env.store->ctx();
-  boost::system::error_code ec;
+class Connection {
+  RGWProcessEnv& env;
+  boost::asio::strand strand;
+  tcp::socket socket;
 
-  beast::flat_streambuf buffer{1024};
+  // references are bound to callbacks for async operations. if a callback
+  // function returns without issuing another operation, the reference is
+  // dropped and the Connection is deleted/closed
+  std::atomic<int> nref{0};
+  using Ref = boost::intrusive_ptr<Connection>;
+
+  // limit header to 4k, since we read it all into a single flat_buffer
+  static constexpr size_t header_limit = 4096;
+  // don't impose a limit on the body, since we read it in pieces
+  static constexpr size_t body_limit = std::numeric_limits<size_t>::max();
+
+  beast::flat_buffer buffer{header_limit};
+  boost::optional<rgw::asio::parser_type> parser;
+
+  std::array<char, 1024> discard_buffer; //< for discard_unread_message()
```

Comment:
Do we need to have the `discard_buffer` as a part of the `Connection`'s state? Are we interested in its content?

Reply:
we never read its content, no. the only requirement is that the buffer stays alive between `discard_unread_message()` and the `on_discard_unread()` callback. keeping it in the connection avoids the need to allocate/free

Reply:
I see. Maybe we could have it as a `static` in `discard_unread_message()`? What do you think?

Reply:
good idea! updated, with a comment explaining that it's okay for multiple clients to share the static discard_buffer